### PR TITLE
Analyze multistage

### PR DESCRIPTION
**Pull request by ForgetMe17**
The commit `Draft version of analyze multistage dockerfile` is a draft implementation of analyzing multistage dockfile.  Here are the steps.
1. split the dockfile for build.
2. build the images.
3. split the dockerfile by stage for analyze.
4. analyze each image and generate a report.

### CODE REVIEW

**Comment by ForgetMe17** [comment]
I have tested on this dockerfile:
```
FROM golang:1.7.3
WORKDIR /go/src/github.com/alexellis/href-counter/

FROM alpine:latest
WORKDIR /root/
```
Here are the html reports by stage.
[output.zip](https://github.com/tern-tools/tern/files/5069888/output.zip)

**Comment by ForgetMe17** [comment]
Hi @abhaykatheria , I have ignored this case.  The dockerfile I used for test did not contain COPY command. This should be a trouble here. Thanks for finding this! I will try to come up with a plan to fix this issue.

**Comment by nishakm** [comment]
@ForgetMe17 I'm holding off merging this PR after a release. Don't worry, I will merge it after cutting the release :)

File: tern/analyze/docker/run.py
Comment (nishakm) [inline tern/analyze/docker/run.py:198]: This is a hard coded path. Also, why the `.html` file extension?

**Review by nishakm — CHANGES_REQUESTED** [review]
@ForgetMe17 Sorry for getting back so late! I have a couple of questions about the code. I think it's reasonably straightforward to build and analyze each stage from here on out.

File: tern/analyze/docker/dockerfile.py
Comment (ForgetMe17) [reply tern/analyze/docker/dockerfile.py:380]: `get_multistage_image_dockerfiles(dfobj_multi)` split the dockerfile like this:
File 1: Stage 1
File 2: Stage 2
and  `split_multistage_dockerfile_by_stage(dfobj_multi)` split the dockerfile like this:
File 1: Stage 1
File 2: Stage 1 + Stage 2

File: tern/analyze/docker/run.py
Comment (ForgetMe17) [reply tern/analyze/docker/run.py:198]: This is a draft version, `html` is easy for read.

**Comment by ForgetMe17** [comment]
Hi @nishakm , I think that i need to consider on the COPY command, there need to be a plan on this.
And for now, I have run this code on some simple dockerfiles(without COPY command) and the result looks good. 
So once we can successfully build the images, we can follow the logic in this PR to analyze.

File: tern/analyze/docker/dockerfile.py
Comment (nishakm) [reply tern/analyze/docker/dockerfile.py:380]: Hmm...but you write out the Dockerfiles appropriately in `get_multistage_image_dockerfiles`. In path `dftemp/0` exists Stage1 and the original Dockerfile is `Stage1+Stage2`.

I tried this with https://github.com/anuvu/zot and it works (at least the Dockerfiles are split into the appropriate stages)

One thing to note is that when you are building with docker, you will have to provide the absolute path to the Dockerfile *in the locally cloned repository*. That should solve the COPY issue you are having.

**Comment by nishakm** [comment]
> Hi @nishakm , I think that i need to consider on the COPY command, there need to be a plan on this.
> And for now, I have run this code on some simple dockerfiles(without COPY command) and the result looks good.
> So once we can successfully build the images, we can follow the logic in this PR to analyze.

COPY only works if there are files to copy. Typically, the Dockerfile will be included in the github repo. So for testing, the github repo needed to be cloned and then you can analyze the Dockerfile.

I think your second commit doesn't need a new splitting function. You do need a new "multistage build and analyze" function in `tern/analyze/docker/analyze.py`. Here you will need to 1. use the multistage split function to make individual dockerfiles 2. For each dockerfile a. build b. analyze.
The final think we ought to think about is how to represent reports for multistage docker builds. For now we will just name the images "imagename_stagex:tag" and then list all the images for the default, json and yaml format. Don't worry about the SPDX format for now.

**Comment by ForgetMe17** [comment]
Hi @nishakm , I am following the steps below in the new commit:
1. split the multistage dockerfile. (`get_multistage_image_dockerfiles()`).
2. iterate on each splited dockerfile by `execute_dockerfile(args)`.
Please take a look on this. If this works fine, i will try to merge this function into `__main__.py` to enable this in the tern project.

To test this commit, i just put execute_multistage_dockerfile(args) in the `__main__.py`:
```
def do_main(args):
    '''Execute according to subcommands'''
    # Set mount/working dir according to user
    rootfs.set_working_dir(args.working_dir)
    # create working directory
    create_top_dir(args.working_dir)
    if not args.quiet:
        # set up console logs
        global logger
        global console
        logger.addHandler(console)
    logger.debug('Starting...')
    if args.clear_cache:
        logger.debug('Clearing cache...')
        cache.clear()
    if hasattr(args, 'name') and (args.name == 'report' or
                                  args.name == 'lock'):
        ***execute_multistage_dockerfile(args)*** HERE!
        if args.name == 'lock':
```

**Comment by nishakm** [comment]
> Hi @nishakm , I am following the steps below in the new commit:
> 
>     1. split the multistage dockerfile. (`get_multistage_image_dockerfiles()`).
> 
>     2. iterate on each splited dockerfile by `execute_dockerfile(args)`.
>        Please take a look on this. If this works fine, i will try to merge this function into `__main__.py` to enable this in the tern project.
> 
> 
> To test this commit, i just put execute_multistage_dockerfile(args) in the `__main__.py`:
> 
> ```
> def do_main(args):
>     '''Execute according to subcommands'''
>     # Set mount/working dir according to user
>     rootfs.set_working_dir(args.working_dir)
>     # create working directory
>     create_top_dir(args.working_dir)
>     if not args.quiet:
>         # set up console logs
>         global logger
>         global console
>         logger.addHandler(console)
>     logger.debug('Starting...')
>     if args.clear_cache:
>         logger.debug('Clearing cache...')
>         cache.clear()
>     if hasattr(args, 'name') and (args.name == 'report' or
>                                   args.name == 'lock'):
>         ***execute_multistage_dockerfile(args)*** HERE!
>         if args.name == 'lock':
> ```

Hmm...I've noticed a few problems and I am proposing some solutions here. Let me know if I am missing anything:
1. I forgot that in order for `COPY` to be reliable, the `stage1` and `stage1+stage2` dockerfiles need to exist in the same directory as the original Dockerfile. So `dftemp` is not a good idea. Instead, just create a `stage1`, `stage12`, etc file names in the same directory as the original Dockerfile (they don't have to be called `Dockerfile`, that's not a requirement by docker).
2. In the default report, we need to add the analysis of each of the stages to the final report. I would say for each stage, make a DockerImage object.
3. We need to clean up the individual stage dockerfiles once we are done with analysis. This is part of teardown.

**Comment by nishakm** [comment]
I am closing this as I have merged it into my working branch. Thanks for your work towards this @ForgetMe17!

**Pull request closed without merging**
